core/controllers/suggestion_test.py

```diff
@@ -188,6 +190,209 @@ def test_create_suggestion(self):
         self.assertEqual(len(suggestions), 3)
         self.logout()
 
+    def test_suggestion_to_exploration_handler_with_invalid_suggestion_id(self):
+        self.login(self.EDITOR_EMAIL)
+
+        response = self.get_html_response('/explore/%s' % self.EXP_ID)
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        suggestion_to_accept = self.get_json(
+            '%s?author_id=%s' % (
+                feconf.SUGGESTION_LIST_URL_PREFIX,
+                self.author_id))['suggestions'][0]
+
+        response = self.get_html_response('/explore/%s' % self.EXP_ID)
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        # Invalid format of suggestion id.
+        response = self.put_json(
+            '%s/exploration/%s/%s' % (
+                feconf.SUGGESTION_ACTION_URL_PREFIX,
+                suggestion_to_accept['target_id'], 'invalid_suggestion_id'),
+            {}, csrf_token=csrf_token, expected_status_int=400)
+
+        self.assertEqual(
+            response['error'],
+            'Invalid format for suggestion_id. It must contain 3 parts '
+            'separated by \'.\'')
+
+        response = self.get_html_response('/explore/%s' % self.EXP_ID)
+        csrf_token = self.get_csrf_token_from_response(response)
+
+        # Suggestion does not exist.
+        self.put_json(
+            '%s/exploration/%s/%s' % (
+                feconf.SUGGESTION_ACTION_URL_PREFIX,
+                suggestion_to_accept['target_id'],
+                'exploration.target_id.id'), {}, csrf_token=csrf_token,
+            expected_status_int=404)
+
+        self.logout()
+
+    def test_suggestion_to_exploration_handler_with_invalid_target_type(self):
+        self.login(self.EDITOR_EMAIL)
+
+        question_dict = {
+            'question_state_data': self._create_valid_question_data(
+                'default_state').to_dict(),
+            'language_code': 'en',
+            'question_state_data_schema_version': (
+                feconf.CURRENT_STATE_SCHEMA_VERSION)
+        }
+
+        thread_id = feedback_services.create_thread(
+            suggestion_models.TARGET_TYPE_QUESTION, self.EXP_ID,
+            self.author_id, 'description', '', has_suggestion=True)
+
+        suggestion_models.GeneralSuggestionModel.create(
+            suggestion_models.SUGGESTION_TYPE_ADD_QUESTION,
+            suggestion_models.TARGET_TYPE_QUESTION, self.EXP_ID,
```

Comment:
This suggestion type is supposed to use suggestion_models.TARGET_TYPE_TOPIC. Don't use the creation functions in the models directly (here and elsewhere). Create the models using the services (which does these checks automatically).

Reply:
There _is_ a service called `create_suggestion` in suggestion_services but that does not allow us to choose the thread id -- it creates one of its own, and we need the thread id to send the request(get_json, post_json etc). The only way to use the service would be to do self.swap but I was confused which was the better choice (creating the model or doing self.swap). Should I use self.swap instead then?

Reply:
And why should the suggestion type necessarily has to be TARGET_TYPE_TOPIC. We are testing it with incorrect target type where the correct type is TARGET_TYPE_EXPLORATION. Hence any other types other than TARGET_TYPE_EXPLORATION should be fine, right?

Reply:
The correct type here is TARGET_TYPE_TOPIC (see suggestion_registry). We have one controller for every type of target. So what you should be testing is that a valid suggestion, passed to the wrong controller should raise an error.

Reply:
You could also get the thread ID by crafting a query to the suggestion list handler (an extra step, but prevents directly manipulating models. By directly adding models, we bypass the validations written in the domain/service layers).

Reply:
Ah yes you are correct. Will fix this!

Reply:
Done